src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,392 @@
+    repo_type: str,
+    repo_id: str,
+    token: str,
+    revision: str,
+    endpoint: Optional[str] = None,
+    num_threads: Optional[int] = 5,
+):
+    """
+    Uploads the content of ``additions`` to the HF Hub using the large file storage protocol.
+
+    Args:
+        additions (``Iterable`` of ``CommitOperationAdd``):
+            The files to be uploaded
+        repo_type (``str``):
+            Type of the repo to upload to: `"model"`, `"dataset"` or `"space"`.
+        repo_id (``str``):
+            A namespace (user or an organization) and a repo name separated
+            by a `/`.
+        token (``str``):
+            An authentication token ( See https://huggingface.co/settings/tokens )
+        revision (``str``):
+            The git revision to upload the files to. Can be any valid git revision.
+        num_threads (``int``, *optional*):
+            The number of concurrent threads to use when uploading. Defaults to 5.
+            If set to `None`, concurrency is disabled.
+
+    """
+    endpoint = endpoint if endpoint is not None else ENDPOINT
+
+    common_headers = {
+        "Accept": "application/vnd.git-lfs+json",
+        "Content-Type": "application/vnd.git-lfs+json",
+    }
+
+    oid2addop = {add_op.upload_info().sha256.hex(): add_op for add_op in additions}
+    batch_url = (
+        f"{base_url(repo_type, repo_id, endpoint=endpoint)}.git/info/lfs/objects/batch"
+    )
+    batch_res = requests.post(
+        batch_url,
+        headers=common_headers,
+        json={
+            "operation": "upload",
+            "transfers": ["basic", "multipart"],
+            "objects": [
+                {
+                    "oid": add_op.upload_info().sha256.hex(),
+                    "size": add_op.upload_info().size,
+                }
+                for add_op in additions
+            ],
+            "ref": {
+                "name": revision,
+            },
+            "hash_algo": "sha256",
+        },
+        auth=HTTPBasicAuth("access_token", token),
+    )
+    batch_res.raise_for_status()
+
+    try:
+        payload: LfsBatchResponse = batch_res.json()
+        objects = payload.get("objects", [])
+        errors: List[LfsBatchObjectError] = [
+            obj for obj in objects if obj.get("error") is not None
+        ]
+        if errors:
+            message = "\n".join(
+                [
+                    f'Encountered error for file with OID {obj["oid"]}:'
+                    f' `{obj["error"]["message"]}'
+                    for obj in errors
+                ]
+            )
+            raise ValueError(f"LFS batch endpoint returned errors:\n{message}")
+
+        upload_func = partial(_upload_lfs_object, oid2addop=oid2addop)
+        if num_threads is None:
+            [upload_func(obj) for obj in objects]
+        else:
+            with ThreadPoolExecutor(num_threads) as pool:
+                wait([pool.submit(upload_func, obj) for obj in objects])
```

Comment:
> I implemented a very naive multi-threaded upload in https://github.com/huggingface/huggingface_hub/commit/05a040e00dd78c9136494c7de76b062d62af111b
> It's not very elegant, so let me know if you want to handle that in a different PR

Hey @SBrandeis, I'm following up on your last message about multi-threading. 

IMHO, it is a good use case for using `asyncio` here which offers more flexibility on how to handle the futures. However, to do so you need `_upload_lfs_object` to be async, which means `upload_lfs_file` requires an implementation with async calls instead of the  `request.put(...)`. Since `upload_lfs_file` seems to be a public API (please confirm), the decision to switch it to an async method is more of a design decision. A solution to use async here could be to have an internal `_async_upload_lfs_file` that does the job and the public `upload_lfs_file` would just be calling this internal one (and you can reuse the async one in commit_api).


Otherwise, staying with a `ThreadPoolExecutor` like in your solution can do the job. I would be more careful about exceptions. For now if a thread throws an exception, it will be silently ignored. What I would do is:
1. Keep a list of your futures:
```py
futures = [pool.submit(upload_func, obj) for obj in objects]
```
2.  Wait on the futures but only until first exception (see https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait)
```py
completed, pending = wait(futures, return_when=ALL_COMPLETED)
```

3. Cancel pending futures (if any). If a future is pending it means another future raised an exception. Here it's just to avoid uploading other big files if you know at least 1 failed. :warning: Problem is that you can't cancel a task that is already running. IDK a good way to stop a running thread (maybe look into https://superfastpython.com/threadpoolexecutor-stop-tasks/ ?)
```py
for future in pending:
    future.cancel()
```

4. Check that all futures are completed and raise exception if necessary. Idk how HF hub is working if some files have been uploaded but not all (should it delete them ?). I guess that's more a question on the backend side than in this lib/sdk ?
```py
for future in futures:
    future.result() # <- will raise if the future threw an exception
```


Hope this can help. I'd be glad to have other opinions about how to properly handle that as well.

Reply:
I really don't know a lot about concurrency in Python (the last time I used it, 2 years ago, I wanted to kill myself)
but two high-level questions:
- shouldn't we use httpx or similar library?
- are we sure we do want to implement concurrent uploads? (some people will be constrained by total local upload bandwidth, i.e. the speedup of concurrency is a bit hard for me to predict)

Reply:
> shouldn't we use httpx or similar library?

a similar library is aiohttp, and is already a dependency of `datasets` (we use it for streaming)

> some people will be constrained by total local upload bandwidth, i.e. the speedup of concurrency is a bit hard for me to predict

True, but it helps if you have thousands of small files for example (e.g. folder of CSV, PNG, WAV etc.)